linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala

```diff
@@ -78,8 +82,20 @@ final class Emitter(config: Emitter.Config) {
   val injectedIRFiles: Seq[IRFile] = PrivateLibHolder.files
 
   def emit(moduleSet: ModuleSet, logger: Logger): Result = {
+    nameCompressor match {
+      case None =>
+        if (state == null)
+          state = new State(Set.empty)
+      case Some(compressor) =>
+        compressor.startRun(moduleSet, logger)
```

Comment:
Rather than this pre-pass, have you considered always using 2 passes and record the names in the first passes call to `allocatedFor` / `allocatedForAncestor`?

This will of course almost certainly be slower, but IIUC, it makes the whole part of the code where we need to keep the identifiers we actually generate in sync just go away.

Reply:
Hum, I had not considered that. It would indeed get rid of the extra `Traverser` that needs to stay in sync.

It does the make the whole thing quite a bit slower, though. On a cold JVM, generating the classes in 10x slower than collecting the names using this traverser. On a warm JVM, the difference is exacerbated, and becomes on the order of 20x, making the name collection essentially free.

<details>
<summary>Cold testSuite2_12/Test/fullLinkJS</summary>
<pre>
[debug] Linker: Compute reachability: 1128593 us
[debug] Linker: Assemble LinkedClasses: 59189 us
[debug] Linker: Check IR: 228524 us
[debug] Linker: 1433503 us
[debug] Optimizer: Batch mode: true
[debug] Optimizer: Incremental part: 241190 us
[debug] Optimizer: Optimizing 48762 methods.
[debug] Optimizer: Optimizer part: 8157130 us
[debug] Optimizer: 8419829 us
[debug] Refiner: Compute reachability: 500739 us
[debug] Refiner: Assemble LinkedClasses: 33364 us
[debug] Refiner: 536852 us
[debug] Module Splitter: Calculate Dependency Info: 3605 us
[debug] Module Splitter: Analyze Modules: 836 us
[debug] Module Splitter: Assemble Modules: 8974 us
[debug] Module Splitter: 14184 us
[debug] Name compressor: Collect names: 252313 us
[debug] Name compressor: Allocate property names: 20928 us
[debug] Name compressor: Allocate ancestor names: 5217 us
[debug] Emitter: Generate Classes: 2017037 us
[debug] Emitter: Write trees: 32278 us
[debug] Emitter: Class tree cache stats: reused: 0 -- invalidated: 3853
[debug] Emitter: Method tree cache stats: reused: 0 -- invalidated: 31970
[debug] Emitter: 2400253 us
[debug] BasicBackend: Write result: 855648 us
[debug] BasicBackend: total top-level trees: 58534; re-computed: 58534
[debug] BasicBackend: total modules: 1; re-written: 1
</pre>
</details>

<details>
<summary>Warm testSuite2_12/Test/fullLinkJS</summary>
<pre>
[debug] Linker: Compute reachability: 168885 us
[debug] Linker: Assemble LinkedClasses: 34929 us
[debug] Linker: Check IR: 266421 us
[debug] Linker: 485198 us
[debug] Optimizer: Batch mode: false
[debug] Optimizer: Incremental part: 106759 us
[debug] Optimizer: Optimizing 0 methods.
[debug] Optimizer: Optimizer part: 53 us
[debug] Optimizer: 124440 us
[debug] Refiner: Compute reachability: 114898 us
[debug] Refiner: Assemble LinkedClasses: 38497 us
[debug] Refiner: 155422 us
[debug] Module Splitter: Calculate Dependency Info: 2455 us
[debug] Module Splitter: Analyze Modules: 372 us
[debug] Module Splitter: Assemble Modules: 4072 us
[debug] Module Splitter: 7050 us
[debug] Name compressor: Collect names: 56550 us
[debug] Name compressor: Allocate property names: 10053 us
[debug] Name compressor: Allocate ancestor names: 4293 us
[debug] Emitter: Generate Classes: 1403884 us
[debug] Emitter: Write trees: 8105 us
[debug] Emitter: Class tree cache stats: reused: 0 -- invalidated: 3853
[debug] Emitter: Method tree cache stats: reused: 0 -- invalidated: 31970
[debug] Emitter: 1513240 us
[debug] BasicBackend: Write result: 667349 us
[debug] BasicBackend: total top-level trees: 58534; re-computed: 58534
[debug] BasicBackend: total modules: 1; re-written: 1
</pre>
</details>

Reply:
I can give it a shot later this week. Probably on Tuesday.

Reply:
Yeah, those numbers do not look encouraging.

IIUC once thing we do not handle properly with the current approach is to make sure we do not allocate names for things we do not actually need. (IIUC we'll detect the opposite case, since we won't find the compressed name and fail). So we might need a better strategy to keep these things in sync (funnily enough, it seems to be a very similar problem than ensuring the indentation of nested generated blocks I had in #4917: trivial to do if we are willing to pay runtime overhead, but we're probably not :-/).

Reply:
> IIUC once thing we do not handle properly with the current approach is to make sure we do not allocate names for things we do not actually need.

We could record usage of each name during the class generation pass. Then we check afterward that all allocated names have been used at least once. (probably only in our testing infrastructure).

We could go further and check the exact *count* for every name, but then we have to actually accurately count every single one, instead of sometimes doing a little bit of overapproximation (like with hijacked method names).

Reply:
> Then we check afterward that all allocated names have been used at least once. (probably only in our testing infrastructure).

Yes, I think if we manage to do that without incurring (too much) overhead in normal execution, that could work.

> We could go further and check the exact count for every name, but then we have to actually accurately count every single one, instead of sometimes doing a little bit of overapproximation (like with hijacked method names).

I'm less concerned about the correct counts: It's much easier to spot that an implementation is incorrect than that an implementation is unnecessary.